### ISSUE TITLE
Analyzer test classes don't detect exceptions thrown by analyzers

### ISSUE THREAD

**Issue by lgolding** [dotnet/roslyn#2748]
Repro:
1. Write a diagnostic analyzer that throws an exception, for example:
   
   ```
   public override void Initialize(AnalysisContext context)
   {
       context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.NamespaceDeclaration);
   }
   
   private void AnalyzeNode(SyntaxNodeAnalysisContext context)
   {
       throw new NotImplementedException();
   }
   ```
2. Write a unit test for a code fragment that I do not expect to emit any diagnostics:
   
   ```
   [TestMethod]
   public void TestMethod1()
   {
       var test = @"namespace N {}";
   
       VerifyCSharpDiagnostic(test);
   }
   ```

Expected: The test should fail, and report that the analyzer threw an exception.

Actual: The test passes.

The problem is in the DiagnosticVerifier base class that comes with the unit test project in the "Analyzer with Code Fix (NuGet + VSIX)" project template. The method GetSortedDiagnosticsFromDocuments runs the analyzers on a background thread:

```
            var compilationWithAnalyzers = project.GetCompilationAsync().Result.WithAnalyzers(ImmutableArray.Create(analyzer));
            var diags = compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync().Result;
```

If an exception occurs on the background thread, it is lost.

**Comment by mavasani** [comment]
Note that this will not repro in VS IDE or command line builds, as CompilationWithAnalyzers is not used by either of them.

**Comment by Vannevelj** [comment]
Do you have a suggestion as to how this might be fixed? Can it be done with a change to the testing helpers or does it require modifications in Roslyn internals?

**Comment by sharwell** [comment]
This has since been fixed with the new test library (Microsoft.CodeAnalysis.Testing).
